ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java

```diff
@@ -514,68 +514,79 @@ public void processDepositWithoutCheckingMerkleProof(
-        boolean proof_is_valid =
-            !BLS_VERIFY_DEPOSIT
-                || BLS.verify(pubkey, signing_root, deposit.getData().getSignature());
-        if (!proof_is_valid) {
-          if (deposit instanceof DepositWithIndex) {
-            LOG.debug(
-                "Skipping invalid deposit with index {} and pubkey {}",
-                ((DepositWithIndex) deposit).getIndex(),
-                pubkey);
-          } else {
-            LOG.debug("Skipping invalid deposit with pubkey {}", pubkey);
-          }
-          if (pubKeyToIndexMap != null) {
-            // The validator won't be created so the calculated index won't be correct
-            pubKeyToIndexMap.remove(pubkey);
-          }
-          return;
-        }
+      if (depositSignatureIsValid(deposit, pubkey)) {
+        processNewValidator(state, deposit);
+      } else {
+        handleInvalidDeposit(deposit, pubkey, pubKeyToIndexMap);
       }
+    } else {
+      // This validator already exists, increase their balance
+      validatorsUtil.increaseBalance(
+          state, existingIndex.getAsInt(), deposit.getData().getAmount());
+    }
+  }
 
-      if (pubKeyToIndexMap == null) {
-        LOG.debug("Adding new validator to state: {}", state.getValidators().size());
-      }
-      state.getValidators().append(getValidatorFromDeposit(deposit));
-      state.getBalances().appendElement(amount);
+  private void handleInvalidDeposit(
+      final Deposit deposit,
+      BLSPublicKey pubkey,
+      final Map<BLSPublicKey, Integer> pubKeyToIndexMap) {
+    if (deposit instanceof DepositWithIndex) {
+      LOG.debug(
+          "Skipping invalid deposit with index {} and pubkey {}",
+          ((DepositWithIndex) deposit).getIndex(),
+          pubkey);
     } else {
-      validatorsUtil.increaseBalance(state, existingIndex.getAsInt(), amount);
+      LOG.debug("Skipping invalid deposit with pubkey {}", pubkey);
+    }
+    if (pubKeyToIndexMap != null) {
+      // The validator won't be created so the calculated index won't be correct
+      pubKeyToIndexMap.remove(pubkey);
     }
   }
 
+  private boolean depositSignatureIsValid(final Deposit deposit, BLSPublicKey pubkey) {
+    if (!BLS_VERIFY_DEPOSIT) {
+      return true;
+    }
+
+    final UInt64 amount = deposit.getData().getAmount();
+    final DepositMessage deposit_message =
+        new DepositMessage(pubkey, deposit.getData().getWithdrawal_credentials(), amount);
+    final Bytes32 domain = beaconStateUtil.computeDomain(specConfig.getDomainDeposit());
+    final Bytes signing_root = beaconStateUtil.computeSigningRoot(deposit_message, domain);
+    return BLS.verify(pubkey, signing_root, deposit.getData().getSignature());
+  }
+
+  protected void processNewValidator(final MutableBeaconState state, final Deposit deposit) {
+    LOG.debug("Adding new validator to state: {} + 1 validator(s)", state.getValidators().size());
```

Comment:
What does the `+ 1` in this message mean?  `state.getValidators().size()` when this line executes would be the validator index of the new validator which I think was why it was logged before.  I think I'd probably have:
```suggestion
    LOG.debug("Adding new validator with index {} to state", state.getValidators().size());
```

Though it's only debug level so I'm not too concerned about it.

Reply:
Oh, I was reading that as size before, which seemed confusing because it wasn't clear whether the count was before or after incrementing.  Make much more sense as an index 👍.